File: sw/rust/sec/src/proxy.rs
//! Proxy: composes the four modules behind trait objects and switches
//! between the legacy and tagma-sec stacks. The scenario workflow is written
//! only against the trait objects, so the same requirements and outputs are
//! exercised against both implementations.

use crate::audit::{Audit, ChainedAudit};
use crate::authority::Authority;
use crate::channel::{Channel, MacChannel};
use crate::delos::{DelosAuthority, DelosIntegrity};
use crate::integrity::Integrity;
use crate::legacy::{LegacyAuthority, LegacyIntegrity};
use crate::types::{Attestation, Decision, Epoch, Event, Path};

/// A composable security stack with switchable implementations.
pub struct SecStack {
    pub authority: Box<dyn Authority>,
    pub integrity: Box<dyn Integrity>,
    pub audit: Box<dyn Audit>,
    pub channel: Box<dyn Channel>,
}

impl SecStack {
    /// The legacy security pattern stack.
    pub fn legacy() -> Self {
        Self {
            authority: Box::new(LegacyAuthority::new()),
            integrity: Box::new(LegacyIntegrity),
            audit: Box::new(ChainedAudit::new()),
            channel: Box::new(MacChannel),
        }
    }

    /// The tagma-sec pattern stack.
    pub fn delos() -> Self {
        Self {
            authority: Box::new(DelosAuthority::new()),
            integrity: Box::new(DelosIntegrity),
            audit: Box::new(ChainedAudit::new()),
            channel: Box::new(MacChannel),
        }
    }
}

/// Scenario workflow: a coordination node submits a route update.
///
/// The update is accepted only when the attestation authorizes the path at
/// the current epoch, the sealed record verifies, and the event is appended
/// to the audit log. Returns the audit event on success, or `None` on any
/// rejected step.
pub fn route_update(
    stack: &mut SecStack,
    att: &Attestation,
    path: &Path,
    record: &[u8],
    epoch: Epoch,
) -> Option<Event> {
    if stack.authority.authorize(att, path, epoch) != Decision::Allow {
        return None;
    }
    let seal = stack.integrity.seal(record, path, att.principal, epoch);
    if !stack
        .integrity
        .verify(record, path, att.principal, epoch, &seal)
    {
        return None;
    }
    Some(stack.audit.append(record, epoch))
}

File: sw/cpp/tagma_sec/src/proxy.cpp
// Proxy composition and the route-update workflow. Mirrors
// sw/rust/sec/src/proxy.rs.

#include "tagma_sec/proxy.h"

#include "tagma_sec/delos.h"
#include "tagma_sec/hash.h"
#include "tagma_sec/legacy.h"

#include <memory>

namespace tagma_sec {

SecStack SecStack::legacy() {
  SecStack s;
  s.authority = std::make_unique<LegacyAuthority>();
  s.integrity = std::make_unique<LegacyIntegrity>();
  s.audit = std::make_unique<ChainedAudit>();
  s.channel = std::make_unique<MacChannel>();
  return s;
}

SecStack SecStack::delos() {
  SecStack s;
  s.authority = std::make_unique<DelosAuthority>();
  s.integrity = std::make_unique<DelosIntegrity>();
  s.audit = std::make_unique<ChainedAudit>();
  s.channel = std::make_unique<MacChannel>();
  return s;
}

std::optional<RouteUpdate> route_update(SecStack& stack, const Attestation& att,
                                        const Path& path, const Bytes& record,
                                        Epoch epoch) {
  if (stack.authority->authorize(att, path, ACTION_ROUTE_UPDATE, epoch) !=
      Decision::Allow) {
    return std::nullopt;
  }
  const Seal seal = stack.integrity->seal(record, path, att.principal, epoch);
  if (!stack.integrity->verify(record, path, att.principal, epoch, seal)) {
    return std::nullopt;
  }
  const Event event = stack.audit->append(record, epoch);

  // Non-repudiation: bind the record to its path and epoch, exchange a
  // receipt with the channel, and commit the signed evidence to the audit
  // log (receipts are appended by the caller per docs/spec/tagma-sec.md).
  Bytes evidence;
  evidence.reserve(record.size() + path.size() * 2 + 8);
  evidence.insert(evidence.end(), record.begin(), record.end());
  for (const auto& c : path) {
    const Bytes idx = le16(c.index());
    evidence.insert(evidence.end(), idx.begin(), idx.end());
  }
  const Bytes epoch_bytes = le64(epoch);
  evidence.insert(evidence.end(), epoch_bytes.begin(), epoch_bytes.end());

  const Receipt receipt = stack.channel->exchange(evidence, att.principal, epoch);
  stack.audit->append(receipt.signed_evidence.evidence, epoch);

  return RouteUpdate{event, receipt};
}

}  // namespace tagma_sec
